### PR TITLE
Fix some security issues and add Bandit to our CIs

### PR DESCRIPTION
**Pull request by drammock**
I ran [bandit](https://github.com/PyCQA/bandit) on our codebase today. This addresses some of the HIGH and MEDIUM severity security issues that it found, and adds TODOs for some that aren't addressed in this PR (yet).  Summary:

- wherever we use `hashlib.md5` we're using it to hash a file or Python object (not password or other secret) so I've added `usedforsecurity=False` everywhere
- calls to `pickle.loads` are marked as `#noqa B301` when they occur in tests (in all cases we're loading something that we wrote, so the file is known to be safe)
- the one call to `pickle.loads` that is *not* in a test is marked with a TODO
- TODOs added for replacing various `xml.etree.ElementTree.parse` calls with equivalent calls from [`defusedxml` package](https://github.com/tiran/defusedxml), which is the workaround [suggested in the stdlibrary docs](https://docs.python.org/3/library/xml.html#the-defusedxml-package)
- TODO added for sanitizing user-passed values that are used to build up a command-line call passed to `os.system()` (maxfilter)
- TODO for `_requires_module` using `exec()` --- probably OK as-is since users aren't expected to call `_requires_module` anyway, but there might still be a safer alternative
- one change from `eval` to `ast.literal_eval`

There were 7 high and 21 medium issues detected; a couple were false positives (format string misidentified as SQL), some were not exactly false positives but non-issues (like reading pickles that we wrote, in tests), some were genuine (`exec`, `parse`, `os.system`).  There were also >6000 LOW severity issues, which I didn't even look at yet.

I'd be interested to hear other dev's thoughts on this. In particular, whether we should add additional security scanning as part of our CIs. Some options are:

- [bandit](https://github.com/PyCQA/bandit) [docs](https://bandit.readthedocs.io/en/latest/index.html)
- https://github.com/marketplace/actions/pysa-action
- https://github.com/marketplace/actions/hcl-appscan-codesweep
- https://github.com/globocom/huskyCI
- https://github.com/ZupIT/horusec

I haven't reviewed all of these carefully yet.

*note: the `# noqa BXXX` annotations don't actually work, the right way to do it is with a `[tool.bandit]` entry in `pyproject.toml` ... which we can do if we adopt bandit.

### CODE REVIEW

**Comment by larsoner** [comment]
We should use https://github.com/marketplace/actions/bandit-check or similar in https://github.com/mne-tools/mne-python/blob/82b2e82e42c5167f30df371f3ca7e18b86a2313a/.github/workflows/tests.yml#L17

**Comment by drammock** [comment]
OK with me to add `bandit-check` to CIs. WDYT about adding `defusedxml` as a dependency?

**Comment by larsoner** [comment]
> WDYT about adding defusedxml as a dependency?

Based on the `git grep`:
```
$ git grep xml\.etree
mne/channels/_dig_montage_utils.py:import xml.etree.ElementTree as ElementTree
mne/channels/_standard_montage_utils.py:import xml.etree.ElementTree as ElementTree
mne/io/egi/events.py:from xml.etree.ElementTree import parse
mne/io/nedf/nedf.py:from xml.etree import ElementTree
```
it looks like we more or less just use it for EGI/MFF, nedf, and maybe some dig (possibly also from EGI/MFF). Based on how often people share files for debugging and analysis, it does seem like a good idea to protect against something like `read_raw_egi(fname)` being a security risk. And it being recommended by the main Python standard library folks (and it being licensed under Python Software Foundation License, which is BSD-compatible it seems) makes it seem worthwhile 👍

**Comment by larsoner** [comment]
@drammock let me know if you want help with CIs. But I think it's okay to get the CI running with however much work you want to do so far (maybe you've already done enough?) and add ignores to `pyproject.toml` for now. And we tackle the other ones later after opening an issue tagged for 1.6 reminding us to do it

**Comment by cbrnr** [comment]
I think adding bandit checks to our CI is a great idea!

File: requirements.txt [hunk 1]
```diff
@@ -43,3 +43,4 @@ mne-qt-browser
 darkdetect
 qdarkstyle
 threadpoolctl
+defusedxml
```
Comment (larsoner) [inline requirements.txt:46]: Sholud go in `requirements_base.txt` rather than `requirements`

**Review by larsoner — COMMENTED** [review]
> TODO added for sanitizing user-passed values that are used to build up a command-line call passed to os.system() (maxfilter)

I'm actually inclined to deprecate this instead and remove it after 1.6. We have had `mne.preprocessing.maxwell_filter` for some time which should be more fully featured and it's what we want to support going forward. If people want to use the command-line `maxfilter` from MEGIN I think it's okay nowadays to have them do it from the command line directly -- `mne maxwell_filter` does not actually do much for people, so the workaround for deprecation is actually pretty straightforward -- use `maxfilter` directly with nearly identical args. The only extra thing it does is write out an `origin` and this can be done pretty easily with minimal MNE code already I think.

**Comment by drammock** [comment]
OK this is ready for review. Bandit is now running in our CIs: https://github.com/mne-tools/mne-python/actions/runs/6030474031/job/16362411943?pr=11937

Notes: config options are pretty limited. You can exclude files/folders *for all checks* or you can exclude specific checks *for all files* but there is no way to exclude certain checks only on certain files. Thus a handful of `# nosec` comments in our tests, to mark those particular instances of `pickle.loads` as safe. I also marked a couple `# nosec` things that are either deprecated (so they'll go away soon) or have associated `# TODO` comments, just to get things green.

The exception to the "config is limited" statement is `assert_used` which is apparently important enough to get a special config value (presumably because of pytest). I've gone ahead and configured it to ignore `assert` usage in test files, but note that the CI is also currently configured only to report vulnerabilities of severity MEDIUM or higher, so that config is redundant / impotent at the moment. I left it in because one day we might make the CI pickier and allow it to fail for LOW severity issues (in which case we'll still want to ignore `assert` in pytest tests).

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -26,6 +26,17 @@ jobs:
...
 
+  bandit:
+    name: Bandit
+    needs: style
+    runs-on: ubuntu-latest
+    steps:
+      - uses: davidslusser/actions_python_bandit@v1.0.0
+        with:
+          src: "mne"
+          options: "-c pyproject.toml -ll -r"
```
Comment (drammock) [inline .github/workflows/tests.yml:37]: these options are
- `-c path/to/config/file`
- `-l` reports LOW and above, `-ll` MEDIUM and above, `-lll` HIGH and above
- `-r` recursive (all files under `src` directory)

**Comment by hoechenberger** [comment]
Great work, @drammock!

**Comment by larsoner** [comment]
Looks like bandit is not happy with your kwargs fix for 3.8. Maybe make a `_safe_md5` function that does this stuff and put it in `utils` with its own inline ignore, then use this `_safe_md5` everywhere we currently `hashlib.md5`?

**Comment by drammock** [comment]
@larsoner could you look at the CI errors here? They're assertion errors in the brain tests.

File: mne/viz/_brain/_scraper.py [hunk 3]
```diff
@@ -44,7 +45,13 @@ def __call__(self, block, block_vars, gallery_conf):
...
                                 break
                         assert line.startswith("(") and line.endswith(")")
-                        kwargs.update(eval(f"dict{line}"))
+                        # Parse a=1, b='whatever', c=...
+                        # using literal_eval because it's safer
+                        line = line[1:-1]
+                        for item in line.split(","):
+                            item = item.strip()
+                            key, val = item.split("=", maxsplit=1)
+                            kwargs[key] = literal_eval(val)
```
Comment (larsoner) [inline mne/viz/_brain/_scraper.py:54]: @drammock this was the key... `literal_eval` can't handle `dict(...)`, it needs to be `{'a': 1, ...` or so. I thought about using `re.sub` but that would rely on substituting `=`. The current code relies on substituting/splitting on `,` so if we ever use `,` in `# brain.save_movie(...)` magic comments in our code we'll need to change this.

Typing this out I'm tempted to change this back to `eval`. This parses comments only in our tutorials/examples. And if people want to execute malicious code there... they can just do it directly in the code. So I'm going to revert this and `# noqa` it

File: mne/io/kit/coreg.py [hunk 4]
```diff
@@ -70,8 +70,9 @@ def read_mrk(fname):
     elif fname.suffix == ".txt":
         mrk_points = _read_dig_kit(fname, unit="m")
     elif fname.suffix == ".pickled":
+        # TODO should we still be supporting pickled here?
         with open(fname, "rb") as fid:
-            food = pickle.load(fid)
+            food = pickle.load(fid)  # nosec B301
```
Comment (larsoner) [inline mne/io/kit/coreg.py:75]: I would say no. Let's deprecate and see if people complain and we'll make sure some better stable public interface works (like writing to text or FIF).

**Comment by larsoner** [comment]
@drammock I pushed a couple of fixes for one or two of the remaining TODOs, marking for merge-when-green, thanks!

**Comment by drammock** [comment]
Thanks for getting this one over the finish line @larsoner

**Pull request merged**